src/drivers/px4fmu/fmu.cpp

```diff
@@ -811,45 +809,143 @@ PX4FMU::cycle()
+	case RC_SCAN_SBUS:
+		if (rc_scan_begin == 0) {
+			warnx("fmu: RC_SCAN_SBUS: begin");
+			rc_scan_begin = now;
+//			// Configure serial port for SBUS
+			sbus_config(_rcs_fd, false);
+//			/* for PixRacer R07, this signal is active low */
+//			/* for PixRacer R12, this signal is active high */
+			stm32_gpiowrite(GPIO_SBUS_INV, 0);
+		} else if (hrt_absolute_time() - rc_scan_last_lock < rc_scan_max
+				|| hrt_absolute_time() - rc_scan_begin < rc_scan_max) {
+
+			bool sbus_failsafe, sbus_frame_drop;
+			uint16_t raw_rc_values[input_rc_s::RC_INPUT_MAX_CHANNELS];
+			uint16_t raw_rc_count;
+
+			// read port
+			rc_updated = sbus_input(_rcs_fd, &raw_rc_values[0], &raw_rc_count,
+						       &sbus_failsafe, &sbus_frame_drop,
+						       input_rc_s::RC_INPUT_MAX_CHANNELS);
+			if (rc_updated) {
+				// we have a new SBUS frame. Publish it.
+				_rc_in.channel_count = raw_rc_count;
+
+				if (_rc_in.channel_count > input_rc_s::RC_INPUT_MAX_CHANNELS) {
+					_rc_in.channel_count = input_rc_s::RC_INPUT_MAX_CHANNELS;
+				}
+
+				for (uint8_t i = 0; i < _rc_in.channel_count; i++) {
+					_rc_in.values[i] = raw_rc_values[i];
+				}
 
-	if (sbus_updated) {
-		// we have a new PPM frame. Publish it.
-		_rc_in.channel_count = raw_rc_count;
+				_rc_in.timestamp_publication = hrt_absolute_time();
+				_rc_in.timestamp_last_signal = _rc_in.timestamp_publication;
 
-		if (_rc_in.channel_count > input_rc_s::RC_INPUT_MAX_CHANNELS) {
-			_rc_in.channel_count = input_rc_s::RC_INPUT_MAX_CHANNELS;
-		}
+				_rc_in.rc_ppm_frame_length = 0;
+				_rc_in.rssi = (!sbus_frame_drop) ? RC_INPUT_RSSI_MAX : (RC_INPUT_RSSI_MAX / 2);
+				_rc_in.rc_failsafe = sbus_failsafe;
+				_rc_in.rc_lost = false;
+				_rc_in.rc_lost_frame_count = sbus_dropped_frames();
+				_rc_in.rc_total_frame_count = 0;
 
-		for (uint8_t i = 0; i < _rc_in.channel_count; i++) {
-			_rc_in.values[i] = raw_rc_values[i];
+				rc_scan_last_lock = now;
+				rc_scan_locked = true;
+			}
+		} else if (!rc_scan_locked) {
+			// This triggers the port re-configuration
+			rc_scan_begin = 0;
+			// Scan the next protocol
+			_rc_scan_state = RC_SCAN_DSM;
 		}
+		break;
 
-		_rc_in.timestamp_publication = hrt_absolute_time();
-		_rc_in.timestamp_last_signal = _rc_in.timestamp_publication;
+	case RC_SCAN_DSM:
+		if (rc_scan_begin == 0) {
+			warnx("fmu: RC_SCAN_DSM: begin");
+			rc_scan_begin = now;
+//			// Configure serial port for DSM
+			dsm_config(_rcs_fd);
+//			/* for PixRacer R07, this signal is active low */
+//			/* for PixRacer R12, this signal is active high */
+			stm32_gpiowrite(GPIO_SBUS_INV, 1);
+		} else if (hrt_absolute_time() - rc_scan_last_lock < rc_scan_max
+				|| hrt_absolute_time() - rc_scan_begin < rc_scan_max) {
+
+			uint8_t n_bytes = 0;
+			uint8_t *bytes;
+			bool dsm_11_bit;
+			uint16_t raw_rc_values[input_rc_s::RC_INPUT_MAX_CHANNELS];
+			uint16_t raw_rc_count;
+
+			rc_updated = dsm_input(_rcs_fd, raw_rc_values, &raw_rc_count, &dsm_11_bit, &n_bytes, &bytes,
```

Comment:
It becomes easier if you pull out the serial read and do the dsm_parse here. Then you can copy that pattern for SUMD and ST24 (read and parse) and call it done.